From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Thu, 19 Jul 2018 14:53:34 -0700
Subject: [PATCH] QtQml: disable use of disk caches when running as root

When you run applications as root, you're often trying to figure out
what's wrong with your system (recovery mode), so loading from a
possibly-corrupt cache is one complexity you really don't want.

[ChangeLog][QtQml][Disk caching] Disk caching is now disabled by default
when the application is running as the root user on Unix systems. To
force using the cache even as root, set the QML_FORCE_DISK_CACHE
environment variable to 1.

Change-Id: Id2be776c7ae0467c9d9ffffd1542e3baadb81867
---
 src/qml/qml/qqmltypeloader.cpp | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/qml/qml/qqmltypeloader.cpp b/src/qml/qml/qqmltypeloader.cpp
index 1d66e756..9acc74f3 100644
--- a/src/qml/qml/qqmltypeloader.cpp
+++ b/src/qml/qml/qqmltypeloader.cpp
@@ -60,6 +60,7 @@
 #include <QtCore/qthread.h>
 
 #include <functional>
+#include <unistd.h>
 
 // #define DATABLOB_DEBUG
 #ifdef DATABLOB_DEBUG
@@ -68,11 +69,21 @@
 #define ASSERT_LOADTHREAD()
 #endif
 
-DEFINE_BOOL_CONFIG_OPTION(disableDiskCache, QML_DISABLE_DISK_CACHE);
+DEFINE_BOOL_CONFIG_OPTION(dumpErrors, QML_DUMP_ERRORS);
+DEFINE_BOOL_CONFIG_OPTION(disableDiskCacheEnv, QML_DISABLE_DISK_CACHE);
 DEFINE_BOOL_CONFIG_OPTION(forceDiskCache, QML_FORCE_DISK_CACHE);
 
 QT_BEGIN_NAMESPACE
 
+static bool disableDiskCache()
+{
+#ifdef Q_OS_UNIX
+    if (getuid() == 0)
+        return true;
+#endif
+    return disableDiskCacheEnv();
+}
+
 namespace {
 
     template<typename LockType>
